Add check for replacing try-catch-fail with assertThatThrownBy

fawind opened the pull request:
## After this PR

This PR adds an ErrorProne BugChecker to use an AssertJ `assertThatThrownBy`-lambda instead of try-catch-fail patterns.

For example given the test case:
```java
try {
  myThrowingMethod();
  fail("My fail message");
} catch (RuntimeException e) {
  // expected
}
```

The BugChecker will output the fix:
```java
assertThatThrownBy(() -> myThrowingMethod())
        .isInstanceOf(RuntimeException.class)
        .describedAs("My fail message");
```

In its current version, we only focus on simple try-catch-fail patterns. Meaning we skip cases that have:
* Multiple catch or a finally block.
* No or only a `fail()` statement in the try block.
* An empty try block.
* A catch block that is not empty.
* try-catch statements that are not in a test class.

## Possible downsides?

Currently we are not checking if the code inside the `try` block references outside code. In this case, the transformation is not correct because the variable is not final but referenced from inside a lambda.

```java
Iterator<String> iter = ...;
try {
  iter.next();
  fail("fail");
} catch (NoSuchElementException e) {}
```

fawind committed:
Set source compatibility to 1.8

fawind committed:
Add generated changelog entries

fawind committed:
Add automation for replacing try-catch-fail with assertThatThrownBy

iamdanfox commented on assertj-error-prone/src/main/java/com/palantir/assertj/errorprone/AssertjAssertThatThrownBy.java:
```diff
@@ -0,0 +1,118 @@
+        if (!catchTree.getBlock().getStatements().isEmpty()
+                || catchTree.getParameter().getType().getKind() == Tree.Kind.UNION_TYPE) {
+            return Description.NO_MATCH;
+        }
```
if we decide we want to handle more complicated catches, I think a reasonable assertj formulation could be:

```java
assertThatThrownBy(() -> {
                    doStuff();
                })
                .isInstanceOfAny(IllegalArgumentException.class, IllegalStateException.class);
```

Don't think we need to do this _just_ yet though, as I haven't seen a ton of these in the wild

fawind replied:
That looks good! Happy to work on this in a follow up!

carterkozak commented on assertj-error-prone/src/main/java/com/palantir/assertj/errorprone/AssertjAssertThatThrownBy.java:
```diff
@@ -0,0 +1,118 @@
+            return Description.NO_MATCH;
+        }
+        CatchTree catchTree = Iterables.getOnlyElement(tree.getCatches());
+        if (!catchTree.getBlock().getStatements().isEmpty()
```
In a future improvement we can handle non-empty blocks using `isInstanceOfSatisfying(type, catch-block-as-lambda)`

carterkozak commented on assertj-error-prone/src/main/java/com/palantir/assertj/errorprone/AssertjAssertThatThrownBy.java:
```diff
@@ -0,0 +1,118 @@
+        return Optional.ofNullable(Iterables.get(failArgs, 0, null))
+                .map(state::getSourceForNode)
+                .filter(msg -> !IGNORED_FAIL_MESSAGES.contains(msg));
+    }
```
We could probably simplify this and ignore the fail message entirely. In most cases it just tells us the same thing as the dynamic assertj failure. Thoughts?

iamdanfox replied:
I dunno, I kinda prefer for our refactors to preserve user-authored stuff where possible, as i think we lose trust in the very rare cases where we delete something that was actually important... e.g. 

```java
       try {
            foo.getPhoneNumber();
            fail("Getting the phone number of an interface without an annotation should fail");
        } catch (IllegalArgumentException e) {

        }
```

carterkozak replied:
That’s a good point :+1:

carterkozak commented on assertj-error-prone/src/test/java/com/palantir/assertj/errorprone/AssertjAssertThatThrownByTest.java:
```diff
@@ -0,0 +1,263 @@
+                        "      System.out.println();",
+                        "    })",
+                        "    .isInstanceOf(RuntimeException.class)",
+                        "    .describedAs(\"My error message.\");",
```
I don't think this actually works (and will fail another error-prone check for ignored return values). The describedAs block must occur prior to checks, which fail immediately when they're executed.

fawind replied:
Had problems reproducing this. It might only apply for a newer version of AssertJ? Anyway, I switched the order of the `.describedAs()` and `.isInstanceOf()` calls so that `describedAs()` comes first.

carterkozak commented on assertj-error-prone/src/main/java/com/palantir/assertj/errorprone/AssertjAssertThatThrownBy.java:
```diff
@@ -0,0 +1,118 @@
+        fix.replace(
+                ((JCTree) tree).getStartPosition(),
+                ((JCTree) throwingStatements.iterator().next()).getStartPosition(),
+                "assertThatThrownBy(() -> {");
```
Do you think we could special case blocks reduced to a single line and leave out the curly braces?

fawind replied:
This is already done by the formatter which runs after all the fixes. I would prefer not adding this logic in code and leave it up for the formatter.

carterkozak replied:
I don't think our formatter knows how to convert between statement and expression lambdas because there's potential for ambiguity.

fawind replied:
You are right! I added the check if the try-block contains only a single statement (not counting the `fail()`). In this case, the single-line lambda is now used. Thanks for flagging!

carterkozak reviewed (commented):
Generally looks good, I'm excited to see this roll out!

ferozco commented:
As an extension (probably in a follow up) we should be able to convert cases where the catch block is not empty to use `assertThatThrownBy(..).isInstanceOfSatisfying(..)`

fawind committed:
Fix order of isInstanceOf and describedAs calls

fawind committed:
Add comments to tree replacements

fawind committed:
Use expression lambda for single method try-blocks

fawind commented:
@ferozco: That's a good idea! Happy to do this in a follow up to keep the first run simple!

@all: Added support for single-line lambdas if there is only one statement in the try-block (without the `fail`). Doing all the replacements with the `SuggestedFix.Builder` got quite messy. I switched it to building the whole `assertThatThrownBy` block first and then replacing the complete try-catch-statement with it. Imo it's also a bit easier to follow. Let me know what you think!

svc-autorelease commented:
Released [0.1.5](https://github.com/palantir/assertj-automation/releases/tag/0.1.5)
